### PR TITLE
Update FOSCKEditor dependency to 2.0

### LINKED ISSUE #447: Upgrade FOS CKEditor dependency to 2.0
## Feature Request

Is it possible to update the **friendsofsymfony/ckeditor-bundle** to 2.0 ?
To fix some deprecation in Symfony and prepare bundle to 5.0 compatibility

### PR DESCRIPTION
**Pull request by psyray**
## Subject

This PR update dependency to FOSCKEditor v2.0 to remove a lot of deprecation warnings in SF 4 and prepare this bundle to SF 5

I am targeting this branch, because it's the main branch

Closes #447 

## Changelog

```markdown
### Changed
Change form types to use the main fos_ck_editor.configuration service

### Removed
Deprecated FOS CKEditor services

### Fixed
Some SF deprecation warnings (~50)

```

## To do
    
- [ ] Update the tests
- [ ] More functional tests
- [ ] Add an upgrade note

In my project, CKEditor is loading fine, changing Toolbar works well.
Assets need to be refreshed.
I need more testing & feedback

### CODE REVIEW

**Comment by psyray** [comment]
@greg0ire @soullivaneuh @OskarStark 
It's a work in progress,
I'm waiting for your comments ;)

File: src/Form/Type/FormatterType.php [hunk 1]
```diff
@@ -44,39 +41,18 @@ final class FormatterType extends AbstractType
...
-     * @var TemplateManagerInterface
-     */
-    private $templateManager;
-
-    /**
-     * @var ToolbarManagerInterface
-     */
-    private $toolbarManager;
-
+    protected $ckEditorConfiguration;
```
Comment (phansys) [inline src/Form/Type/FormatterType.php]: I guess this doesn't need to be protected since the class is declared as final.
```suggestion
    private $ckEditorConfiguration;
```

File: src/Form/Type/FormatterType.php [hunk 2]
```diff
@@ -156,37 +132,32 @@ public function buildView(FormView $view, FormInterface $form, array $options):
...
-        if ($this->configManager->hasConfig($defaultConfig)) {
-            $ckeditorConfiguration = $this->configManager->getConfig($defaultConfig);
-        } else {
+        try {
+            $ckeditorConfiguration = $this->ckEditorConfiguration->getConfig($defaultConfig);
+        } catch (Exception $e) {
             $ckeditorConfiguration = [];
         }
-
+
```
Comment (phansys) [inline src/Form/Type/FormatterType.php]: Please, revert this change.

File: src/Form/Type/FormatterType.php [hunk 3]
```diff
@@ -44,39 +41,18 @@ final class FormatterType extends AbstractType
...
-     * @var ToolbarManagerInterface
-     */
-    private $toolbarManager;
-
+    protected $ckEditorConfiguration;
+    
     public function __construct(
         PoolInterface $pool,
         TranslatorInterface $translator,
-        ConfigManagerInterface $configManager,
```
Comment (core23) [inline src/Form/Type/FormatterType.php]: Technically this is a BC break, also the class is final :/

Comment (psyray) [reply src/Form/Type/FormatterType.php]: Corrected, you reviewed too quickly :)

File: src/Form/Type/FormatterType.php (hunk 1 above)
Comment (psyray) [reply src/Form/Type/FormatterType.php]: done

File: src/Form/Type/FormatterType.php (hunk 2 above)
Comment (psyray) [reply src/Form/Type/FormatterType.php]: done

**Comment by psyray** [comment]
> I guess there is no way to allow versions 1 and 2, don't?

I don't think so

**Comment by psyray** [comment]
@OskarStark 
PHPStan seem to have problem
image

**Comment by OskarStark** [comment]
You can ignore this PHPStan error for now

File: src/Form/Type/FormatterType.php [hunk 4]
```diff
@@ -156,11 +132,11 @@ public function buildView(FormView $view, FormInterface $form, array $options):
...
 
-        $defaultConfig = $this->configManager->getDefaultConfig();
+        $defaultConfig = $this->ckEditorConfiguration->getDefaultConfig();
 
-        if ($this->configManager->hasConfig($defaultConfig)) {
-            $ckeditorConfiguration = $this->configManager->getConfig($defaultConfig);
-        } else {
+        try {
+            $ckeditorConfiguration = $this->ckEditorConfiguration->getConfig($defaultConfig);
+        } catch (Exception $e) {
```
Comment (phansys) [inline src/Form/Type/FormatterType.php]: Could we trust a more specific exception?

Comment (psyray) [reply src/Form/Type/FormatterType.php]: Like `FOSCKEditorException` ?
You want me to throw an exception instead of sending an empty array ?
Like
```
throw new \RuntimeException('Unable to find CKEditor default config, check your fos_ck_editor.yaml file in your config folder, error is: '.$e->getMessage(), $e->getCode(), $e);
```

Comment (phansys) [reply src/Form/Type/FormatterType.php]: I don't know what exceptions are able to be thrown from `getConfig()` method, but I think the exception catched here must be narrow enough in order to avoid silencing other unrelated potential exceptions.

**Comment by wbloszyk** [comment]
> > I guess there is no way to allow versions 1 and 2, don't?
> 
> I don't think so

It is possible by checking interface and include correct file version (look at SonataUserExtension.php). In FormatterType check instance in constructor. I know it is not the best practice but it will be prepare bundle for next major. (feature compatybility?)

In next major version 1 will be drop.

File: src/Form/Type/FormatterType.php (hunk 4 above)
Comment (timwentzell) [reply src/Form/Type/FormatterType.php:139]: Hey guys, hope this hasn't gone stale, just my thoughts - the getConfig() method throws a ConfigException::configDoesNotExist exception when the config requested is not set. I don't see why you would want to catch that exception and ignore it by setting the the editor `$ckeditorConfiguration` to an empty array. I think it would be best to just let the FOSCKEditorBundle handle throwing the exception if an expect config does not exist (since that indicates an issue with the bundle configuration).

Comment (psyray) [reply src/Form/Type/FormatterType.php:139]: @timwentzell Thanks for the tip
You're right, it's better to let FOSCkEditorBundle handle the error.
I will correct this, and also look at the @wbloszyk comment below

**Comment by haivala** [comment]
I hope this gets merged soon!

**Comment by core23** [comment]
> I hope this gets merged soon!

We can't merge his unless someone is fixing the open issues in this PR.

**Comment by core23** [comment]
Looks like you have done a rebase against the wrong branch @psyray

**Comment by psyray** [comment]
PR is rebased & fixed.

@phansys According to @timwentzell suggestion Exception catching has been removed to let FOSCkEditorBundle handle it.

@wbloszyk I check your suggestion, but I don't think it's good. I don't like this bad hack into code.
Maybe we should wait major, and 1.0 removal, to merge this PR.
I don't want to spend time coding this thing, sorry.

Waiting that day, for those who wants to get my work in their project, just fork my repository, and change your composer.json to call your repository
by using
```
    "repositories": [
        {
            "type" : "vcs",
            "url" : "https://github.com/YOUR_NICK/SonataFormatterBundle"
        }
    ],
```
After you have 2 solutions, 
- Create a release higher than the current one (eg: 4.99), you don't need to change version in the **require** line
- Change the version to `dev-master` in the *require* line

And do a composer update

Source : https://getcomposer.org/doc/05-repositories.md#using-private-repositories

**Comment by psyray** [comment]
> Looks like you have done a rebase against the wrong branch @psyray

I have rebased on 4.x, no ?

**Comment by core23** [comment]
> > Looks like you have done a rebase against the wrong branch @psyray
> 
> I have rebased on 4.x, no ?

When I look at the commits of this PR, there are some unrelated commits in it. Please verify if you used the upstream repository and not your forked repository.

> Maybe should I create another PR ?

That would not solve the problem :|

**Comment by psyray** [comment]
@core23 
I don't really know how to fix the situation...
An idea ?

**Comment by psyray** [comment]
Closed due to bad rebase
See #467 for new PR
